compiler/rustc_lint/src/autorefs.rs

```diff
@@ -0,0 +1,180 @@
+use rustc_ast::{BorrowKind, UnOp};
+use rustc_hir::{Expr, ExprKind, Mutability};
+use rustc_middle::ty::{
+    adjustment::{Adjust, Adjustment, AutoBorrow, OverloadedDeref},
+    TyCtxt, TypeckResults,
+};
+use rustc_session::declare_lint;
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+use crate::{
+    lints::{ImplicitUnsafeAutorefsDiag, ImplicitUnsafeAutorefsSuggestion},
+    LateContext, LateLintPass, LintContext,
+};
+
+declare_lint! {
+    /// The `dangerous_implicit_autorefs` lint checks for implicitly taken references
+    /// to dereferences of raw pointers.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// use std::ptr::addr_of_mut;
+    ///
+    /// unsafe fn fun(ptr: *mut [u8]) -> *mut [u8] {
+    ///     addr_of_mut!((*ptr)[..16])
+    ///     //                 ^^^^^^ this calls `IndexMut::index_mut(&mut ..., ..16)`,
+    ///     //                        implicitly creating a reference
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// When working with raw pointers it's usually undesirable to create references,
+    /// since they inflict a lot of safety requirement. Unfortunately, it's possible
+    /// to take a reference to a dereference of a raw pointer implicitly, which inflicts
+    /// the usual reference requirements without you even knowing that.
```

Comment:
```suggestion
/// the usual reference requirements potentially without you realizing that.
```

More neutral, less “condescending” imo.